### DISCUSSION TITLE
No errors on submitting nested subtransactions but the changes does not take effect

### DISCUSSION THREAD

**Discussion by Hooollin** [sqlalchemy/sqlalchemy#11839] [Usage Questions]
I'm using SQLAlchemy 1.2 to integrate with SQLite for running Openstack Cinder unit tests. I found that when submitting nested subtransactions, there are no errors thrown, but the updates to the data are not successful. 
The pseudocode as follows:
```python
# same session is used in volume_update() and backup_update()
# if we update volume first transaction will work. 
# I have identified that the issue is caused by the nested transaction 
# within volume_user_metadata_update(), but I don't know why it behaves this way.
with session.begin():
    backup.state = "restoring"
    db.backup_update({"status": "restoring"})
    db.volume_update(
    {"status": "restoring-backup",
    "metadata": {"restoring_progress": 0}}
    )

def volume_update(context, volume_id, values):
    session, sub_txn = get_or_create_session(context)
    with session.begin(subtransactions=sub_txn):
        metadata = values.get('metadata')
        if metadata is not None:
            _volume_user_metadata_update(context,
                                         volume_id,
                                         values.pop('metadata'),
                                         delete=True,
                                         session=session)

        admin_metadata = values.get('admin_metadata')
        if is_admin_context(context) and admin_metadata is not None:
            _volume_admin_metadata_update(context,
                                          volume_id,
                                          values.pop('admin_metadata'),
                                          delete=True,
                                          session=session)

        query = _volume_get_query(context, session, joined_load=False)
        result = query.filter_by(id=volume_id).update(values)
        if not result:
            raise exception.VolumeNotFound(volume_id=volume_id)


def _volume_user_metadata_update(context, volume_id, metadata, delete,
                                 session=None):
    return _volume_x_metadata_update(context, volume_id, metadata, delete,
                                     models.VolumeMetadata,
                                     session=session)

def _volume_x_metadata_update(context, volume_id, metadata, delete, model,
                              session=None, add=True, update=True):
    session = session or get_session()
    metadata = metadata.copy()

    with session.begin(subtransactions=True):
        # Set existing metadata to deleted if delete argument is True.  This is
        # committed immediately to the DB
        if delete:
            expected_values = {'volume_id': volume_id}
            # We don't want to delete keys we are going to update
            if metadata:
                expected_values['key'] = db.Not(metadata.keys())
            conditional_update(context, model,
                               {'deleted': True,
                                'deleted_at': timeutils.utcnow()},
                               expected_values)

        # Get existing metadata
        db_meta = _volume_x_metadata_get_query(context, volume_id, model).all()
        save = []
        skip = []

        # We only want to send changed metadata.
        for row in db_meta:
            if row.key in metadata:
                value = metadata.pop(row.key)
                if row.value != value and update:
                    # ORM objects will not be saved until we do the bulk save
                    row.value = value
                    save.append(row)
                    continue
            skip.append(row)

        # We also want to save non-existent metadata
        if add:
            save.extend(model(key=key, value=value, volume_id=volume_id)
                        for key, value in metadata.items())
        # Do a bulk save
        if save:
            session.bulk_save_objects(save, update_changed_only=True)

        # Construct result dictionary with current metadata
        save.extend(skip)
        result = {row['key']: row['value'] for row in save}
    return result
```
Any ideas why does that transaction won't take effect? And why do we switch the update sequence it will work again? Thanks in advance.

**Comment by zzzeek** [comment]
hi I work on Openstack for my day job.    Is this part of openstack work on Cinder where you could point me to a gerrit and ongoing efforts?  then again this would be a very old version of Cinder if it's using SQLAlchemy 1.2 and the subtransactions parameter, are you making your own modifications to an old version of Cinder?   transactions being committed or not would be due to your calls being nested inside of a larger transaction, it's not easy to say without looking at the full code in context of a patch.

**Reply by Hooollin** [reply]
I’m indeed working with an older version of OpenStack, specifically the Queens release. Here are the relevant code contexts:
https://github.com/openstack/cinder/blob/queens-eol/cinder/db/sqlalchemy/api.py#L2611
https://github.com/openstack/cinder/blob/queens-eol/cinder/backup/api.py#L394

I’ve initialized a transaction for atomic updates to backup and volume statuses in `backup/api:400` and `backup/api:401`. In SQLAlchemy, both update statements use the same session. However, during unit testing, I’ve observed that the records isn’t updated after the transaction commits. When I swap these two lines of code, the transaction commits and the records is updated as expected. Additionally, removing the `_volume_user_metadata_update()` function from `volume_update()` allows the transaction to commit successfully and updates the records.

It seems to me that nested subtransactions might be causing this issue. Any insights or suggestions?

**Reply by Hooollin** [reply]
Btw, sqlite is the database used in unit test. I did not reproduce the same result using mysql.

**Reply by Hooollin** [reply]
I'm indeed making my own modifications to that version of Cinder. And I just wanna make the status update on backup and volume to be atomic, so I tried to use the same session for update statements.

**Reply by zzzeek** [reply]
you would need to use a new session that is not the same as the old session that uses a distinct connection and transaction

**Reply by Hooollin** [reply]
It seems that using two separate sessions won’t guarantee atomic updates. My goal is to ensure that both the backup and volume status updates are atomic.

**Reply by Hooollin** [reply]
I initialized a session for the 2 update statements. Is that usecase inappropriate?

**Reply by zzzeek** [reply]
you need to locate where the transaction is starting; the commit which matches that level of transaction is the one that commits.    If you are in a big nest of "subtransactions=True" that can be tricky.   I'd run the series of operations with SQL echoing turned on (oslo.db debug level 50 or 100) and see where COMMIT actually shows in the logs.

these patterns are all removed in modern openstack and replaced with oslo.db enginefacade that handles the "nesting" of transactions.   you could still have the same problem but it would be hopefully more explicit where things start and end.

**Reply by Hooollin** [reply]
Here are the levels of transactions:
txn:
--update backup;
--sub_txn:
----sub_txn:
------update volume metadata;
----update volume;

Updates do not take effect with no errors. However, put update backup to the end of the two sub_txns, updates are good.

Thanks a lot for your advice. I’ll try to set log level and provide more useful debug info tomorrow.
